packages/cli/src/models/network/NetworkController.ts

```diff
@@ -313,11 +318,42 @@ export default class NetworkController {
 
   // Contract model || SolidityLib model
   private _getAllSolidityLibNames(contractNames: string[]): string[] {
+    let newLibNames = this._getSolidityLibNames(contractNames);
+    let libNames = newLibNames;
+    while (true) {
+      newLibNames = uniq(this._getSolidityLibNames(newLibNames))
+      
+      // Get new deps
+      const uniqNewLibNames = difference(newLibNames, libNames)
+      
+      // Get repeat deps
+      const repeatLibNames = intersection(newLibNames, libNames)
+
+      // Exclude repeat deps
+      libNames = difference(libNames, repeatLibNames)
+
+      // Prepend deps to ensure they are deployed first
+      const nextLibNames = uniq(uniqNewLibNames.concat(repeatLibNames).concat(libNames))
+
+      // no new libs
+      if (nextLibNames.length === libNames.length) {
+        break;
+      }
+      // otherwise
+      // look into the new libs
+      newLibNames = uniqNewLibNames
+      // update all libs
+      libNames = nextLibNames
+    }
+
+    return libNames;
```

Comment:
May I suggest we use a recursive version? I find it easy to read and understand.
```
    let newLibNames = this._getSolidityLibNames(contractNames);
    // deploy children before parents
    // keep contracts unique
    return uniq([...(newLibNames.length ? this._getAllSolidityLibNames(newLibNames) : []), ...newLibNames]);
```
It does a full walk over a forest though.